File: leetcode/Solution189.java
public class Solution189 {
    public void rotate(int[] nums, int k) {
        if (nums == null || nums.length <= 1) return;

        k %= nums.length;
        reverse(nums, 0, nums.length - k - 1);
        reverse(nums, nums.length - k, nums.length - 1);
        reverse(nums, 0, nums.length - 1);
    }

    private void reverse(int[] nums, int left, int right) {
        while (left < right) {
            nums[left] ^= nums[right];
            nums[right] ^= nums[left];
            nums[left] ^= nums[right];

            left++;
            right--;
        }
    }
}

File: src/leetcode/kotlin/Solution189.kt
package leetcode.kotlin

class Solution189 {
    fun rotate(nums: IntArray, k: Int) {
        val step = k % nums.size
        reverse(nums, 0, nums.size - step - 1)
        reverse(nums, nums.size - step, nums.size - 1)
        reverse(nums, 0, nums.size - 1)
    }

    private fun reverse(nums: IntArray, start: Int, end: Int) {
        var left = start
        var right = end
        while (left < right) {
            nums[left] = nums[right].also { nums[right] = nums[left] }
            left++
            right--
        }
    }
}